test-helpers/src/transfer_aux_data/mod.rs

```diff
@@ -0,0 +1,138 @@
+//! Test helpers for constructing auxiliary data for external transfers
+//! Much of this is ported over from https://github.com/renegade-fi/renegade-contracts/blob/main/integration/src/utils.rs
+
+use alloy_primitives::{keccak256, Address, B256, U256};
+use alloy_sol_types::{
+    eip712_domain,
+    sol_data::{Address as SolAddress, Uint as SolUint},
+    Eip712Domain, SolStruct, SolType,
+};
+use arbitrum_client::{conversion::to_contract_external_transfer, helpers::serialize_calldata};
+use circuit_types::transfers::ExternalTransfer;
+use common::types::transfer_aux_data::{DepositAuxData, TransferAuxData, WithdrawalAuxData};
+use ethers::{signers::Wallet, types::H256};
+use eyre::Result;
+use k256::ecdsa::SigningKey;
+use num_bigint::BigUint;
+use rand::{thread_rng, RngCore};
+
+use self::abi::{PermitTransferFrom, TokenPermissions};
+
+mod abi;
+
+/// The name of the domain separator for Permit2 typed data
+const PERMIT2_EIP712_DOMAIN_NAME: &str = "Permit2";
+
+/// Generates the auxiliary data fpr the given external transfer,
+/// including the Permit2 data & a signature over the transfer
+pub fn gen_transfer_aux_data(
+    wallet: Wallet<SigningKey>,
+    transfer: &ExternalTransfer,
+    permit2_address: Address,
+    darkpool_address: Address,
+    chain_id: u64,
+) -> Result<TransferAuxData> {
+    let transfer = to_contract_external_transfer(transfer)?;
+
+    let res = if transfer.is_withdrawal {
+        let transfer_bytes = serialize_calldata(&transfer)?;
+        let transfer_hash = H256::from_slice(keccak256(&transfer_bytes).as_slice());
+        let transfer_signature = wallet.sign_hash(transfer_hash)?.to_vec();
+
+        TransferAuxData::Withdrawal(WithdrawalAuxData {
+            external_transfer_signature: transfer_signature,
+        })
```

Comment:
still not sure why we have an enum for this? As per my last review it just adds control flow everywhere you want to interact with the struct. Why not just have `TransferAuxData` be a struct and set the deposit fields to `None` on a withdrawal and vice versa? Like how we do it in the contracts

Reply:
Ah, I thought this was ack'd on the last PR - I erred on the side of an enum so that the caller can be sure of exactly what aux data to find properly set, i.e. we don't have to scatter `unwrap`s or `Result`s throughout assuming that all the right fields on the aux data struct are set, this is now enforced by the type system.

I am not horribly opinionated, taking another look at it now, we could have `TransferAuxData` have two fields, an `Option<DepositAuxData>` and an `Option<WithdrawalAuxData>`, which trades off the enum control flow for one `unwrap` / `Result` when accessing the fields. How does that sound?